File: 2217-step-by-step-directions-from-a-binary-tree-node-to-another/step-by-step-directions-from-a-binary-tree-node-to-another.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool findPath(struct TreeNode* node, int value, char* path, int* pathLen) {
    if (node == NULL) {
        return false;
    }
    
    if (node->val == value) {
        return true;
    }
    
    if (findPath(node->left, value, path, pathLen)) {
        path[(*pathLen)++] = 'L';
        return true;
    }
    
    if (findPath(node->right, value, path, pathLen)) {
        path[(*pathLen)++] = 'R';
        return true;
    }
    
    return false;
}

char* getDirections(struct TreeNode* root, int startValue, int destValue) {
    char path1[100000] = {0};
    char path2[100000] = {0};
    int len1 = 0, len2 = 0;
    
    findPath(root, startValue, path1, &len1);
    findPath(root, destValue, path2, &len2);
    
    // Reverse paths
    for (int i = 0; i < len1 / 2; i++) {
        char temp = path1[i];
        path1[i] = path1[len1 - 1 - i];
        path1[len1 - 1 - i] = temp;
    }
    for (int i = 0; i < len2 / 2; i++) {
        char temp = path2[i];
        path2[i] = path2[len2 - 1 - i];
        path2[len2 - 1 - i] = temp;
    }
    
    // Find common prefix
    int i = 0;
    while (i < len1 && i < len2 && path1[i] == path2[i]) {
        i++;
    }
    
    // Construct result
    char* result = (char*)malloc(100000 * sizeof(char));
    int resultLen = 0;
    
    for (int j = i; j < len1; j++) {
        result[resultLen++] = 'U';
    }
    
    for (int j = i; j < len2; j++) {
        result[resultLen++] = path2[j];
    }
    
    result[resultLen] = '\0';
    
    return result;
}

File: 2217-step-by-step-directions-from-a-binary-tree-node-to-another/step-by-step-directions-from-a-binary-tree-node-to-another.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public String getDirections(TreeNode root, int startValue, int destValue) {
        StringBuilder path1 = new StringBuilder();
        StringBuilder path2 = new StringBuilder();
        
        // Find paths from root to start and destination
        findPath(root, startValue, path1);
        findPath(root, destValue, path2);
        
        // Find the common prefix
        int i = 0;
        while (i < path1.length() && i < path2.length() && path1.charAt(i) == path2.charAt(i)) {
            i++;
        }
        
        // Construct the result
        StringBuilder result = new StringBuilder();
        for (int j = i; j < path1.length(); j++) {
            result.append('U');
        }
        result.append(path2.substring(i));
        
        return result.toString();
    }
    
    private boolean findPath(TreeNode node, int value, StringBuilder path) {
        if (node == null) {
            return false;
        }
        
        if (node.val == value) {
            return true;
        }
        
        if (findPath(node.left, value, path)) {
            path.insert(0, 'L');
            return true;
        }
        
        if (findPath(node.right, value, path)) {
            path.insert(0, 'R');
            return true;
        }
        
        return false;
    }
}